fvtest/tril/tril/parser.cpp

```diff
@@ -0,0 +1,482 @@
+        Token(Type type, std::string value):
+	        type(type), value(value) {}
+        Type getType() const { return type; }
+        void setType(Token::Type t) {type = t ;}
+        const std::string &getValue() const {return value ;}
+};
+
+class TokenFailure {
+	std::string message;
+
+public:
+	TokenFailure(std::string message);
+	const std::string &what() const;
+};
+
+TokenFailure::TokenFailure(std::string message) :
+	message(message) {}
+
+const std::string &TokenFailure::what() const { return message; }
+
+int isNotNewLine(int c) { return c != '\n'; }
+
+int isNotQuot(int c) { return c != '"'; }
+
+int isHex(int c) {
+    if (c == 'a' || c == 'b' || c == 'c' || c == 'd' || c == 'e' || c =='f' ||
+        c == 'A' || c == 'B' || c == 'C' || c == 'D' || c == 'E' || c == 'F') {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+int isNotZero(int c) { return c != '0' && isdigit(c); }
+
+class ASTstates {
+    public:
+        enum State {
+            ID = 0,
+            STRING_,
+            STRING_0,
+            INT,
+            WHITESPACE,
+            AT,
+            ZERO,
+            HEXINT,
+            QUOT,
+            PAREN,
+            LPAREN,
+            RPAREN,
+            LPAREN_SQ,
+            RPAREN_SQ,
+            BECOMES,
+            COMMA,
+            DOUBLE_,
+
+            START,
+            FAIL,
+            DOT,
+            MINUS,
+            COMMENT,
+            LARGEST_STATE = COMMENT
+        };
+    
+    private:
+        std::vector<State> acceptingStates;
+        std::vector<std::vector<State> > transitionFunction;
+        Token::Type stateToType(State s) const {
+            switch(s) {
+                case ID: return Token::ID;
+                case STRING_: return Token::STRING_;
+                case INT: return Token::INT;
+                case WHITESPACE: return Token::WHITESPACE;
+                case DOUBLE_: return Token::DOUBLE_;
+                case AT:  return Token::AT;
+                case ZERO:  return Token::ZERO;
+                case QUOT:  return Token::QUOT;
+                case LPAREN:  return Token::LPAREN;
+                case RPAREN:  return Token::RPAREN;
+                case LPAREN_SQ:  return Token::LPAREN_SQ;
+                case RPAREN_SQ:  return Token::RPAREN_SQ;
+                case BECOMES: return Token::BECOMES;
+                case COMMA: return Token::COMMA;
+                case HEXINT:  return Token::HEXINT;
+                case COMMENT: return Token::COMMENT;
+                default: throw TokenFailure("Cannot convert state to Type");
+            }
+        }
+
+    public:
+    
+        State start() const { return START; }
+
+        bool failed(State state) const { return state == FAIL; }
+
+        bool accept(State state) const {
+            for (int i = 0; i < acceptingStates.size(); i++) {
+                if (acceptingStates[i] == state) {
+                    return true;
+                }
+            }
```

Comment:
We can replace this whole function with

```
std::find(acceptingStates.begin(), acceptingStates.end(), state) != acceptingStates.end()
```

In which case all three functions `start`, `failed`, `accept` seem to be one liners which only get used in a single place, so I question their existence.

Reply:
We actually tried this at one point but not all our build compilers seem to support `std::find`.

Reply:
Do you recall which compiler in particular was problematic?

Reply:
GCC 4.8.5 seems to have a problem with `std::find()`:

https://ci.eclipse.org/omr/job/PullRequest-linux_ppc-64_le_gcc/1801/consoleFull